File: HomeAutomation/HomeAutomation/src/main/java/com/homeautomation/core/StateManager.java
package com.homeautomation.core;

import java.util.ArrayList;
import java.util.List;

public class StateManager {
    private final List<Component> components = new ArrayList<>();
    private EnvironmentalState currentState;

    // Register a component to be notified when the state changes
    public void registerComponent(Component component) {
        components.add(component);
    }

    // Unregister a component
    public void unregisterComponent(Component component) {
        components.remove(component);
    }

    // Update the state and notify all registered components
    public void updateState(EnvironmentalState state) {
        this.currentState = state;
        for (Component component : components) {
            component.updateState(state);
        }
    }

    public EnvironmentalState getCurrentState() {
        return currentState;
    }
}

File: HomeAutomation/HomeAutomation/Core/StateManager.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HomeAutomation.Core
{
    public class StateManager
    {
        private readonly List<IComponent> _components = new List<IComponent>();
        public EnvironmentalState CurrentState { get; private set; } = new EnvironmentalState();

        public void RegisterComponent(IComponent component)
        {
            if (!_components.Contains(component))
            {
                _components.Add(component);
            }
        }


        public void UpdateState(Action<EnvironmentalState> updateAction)
        {
            updateAction(CurrentState);
            NotifyComponents();
        }

        private void NotifyComponents()
        {
            foreach (var component in _components)
            {
                component.OnStateChanged(CurrentState);
            }
        }
    }
}
